Twig/Extension/SonataAdminExtension.php

```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
     public function getXEditableType($type)
     {
         $mapping = array(
-            'boolean'    => 'select',
-            'text'       => 'text',
-            'textarea'   => 'textarea',
-            'html'       => 'textarea',
-            'email'      => 'email',
-            'string'     => 'text',
-            'smallint'   => 'text',
-            'bigint'     => 'text',
-            'integer'    => 'number',
-            'decimal'    => 'number',
-            'currency'   => 'number',
-            'percent'    => 'number',
-            'url'        => 'url',
-            'date'       => 'date',
+            'boolean'  => 'select',
+            'text'     => 'text',
+            'textarea' => 'textarea',
+            'html'     => 'textarea',
+            'email'    => 'email',
+            'string'   => 'text',
+            'smallint' => 'text',
+            'bigint'   => 'text',
+            'integer'  => 'number',
+            'decimal'  => 'number',
+            'currency' => 'number',
+            'percent'  => 'number',
+            'url'      => 'url',
+            'date'     => 'date',
         );
 
         return isset($mapping[$type]) ? $mapping[$type] : false;
     }
+
+    /**
+     * Encodes the information needed by Javascript about a form widget to JSON,
+     * to be used in the association handling client-side code.
+     *
+     * @param array $context The current Twig context.
+     *
+     * @return string
+     */
+    public function renderFieldDescriptionToJSON($context)
+    {
+        /** @var AppVariable $app */
+        $app = $context['app'];
+        $request = $app->getRequest();
+        $sonataAdmin = $context['sonata_admin'];
+        /** @var Admin $admin */
+        $admin = $sonataAdmin['admin'];
+        /** @var FieldDescriptionInterface $fieldDescription */
+        $fieldDescription = $sonataAdmin['field_description'];
+        $root = $admin->getRoot();
+        $associationAdmin = $fieldDescription->getAssociationAdmin();
+        $router = $admin->getRouteGenerator();
+
+        $data = array(
+            'id'               => $context['id'],
+            'admin'            => $root->getCode(),
+            'associationAdmin' => $associationAdmin->getCode(),
+            'formType'         => $fieldDescription->getType(),
+            'editMode'         => $sonataAdmin['edit'],
+            'label'            => $associationAdmin->trans($associationAdmin->getLabel(), array(), $associationAdmin->getTranslationDomain()),
```

Comment:
Yeah, about aligned arrays, not sure if alignment is enforced, but if it isn't, just don't do it. Aligning things together is coupling lines that do not related, and makes diffs harder to read.